src/compute-client/src/plan/interpret/physically_monotonic.rs

```diff
@@ -7,5 +7,245 @@
+    /// analysis.
+    pub fn new(monotonic_ids: &'a BTreeSet<GlobalId>) -> Self {
+        SingleTimeMonotonic {
+            monotonic_ids,
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<'a, T> Interpreter<T> for SingleTimeMonotonic<'a, T> {
+    type Domain = PhysicallyMonotonic;
+
+    fn constant(
+        &self,
+        _ctx: &Context<Self::Domain>,
+        rows: &Result<Vec<(Row, T, Diff)>, EvalError>,
+    ) -> Self::Domain {
+        // A constant is physically monotonic iff the constant is an `EvalError`
+        // or all its rows have `Diff` values greater than zero.
+        PhysicallyMonotonic(
+            rows.as_ref()
+                .map_or(true, |rows| rows.iter().all(|(_, _, diff)| *diff > 0)),
+        )
+    }
+
+    fn get(
+        &self,
+        ctx: &Context<Self::Domain>,
+        id: &Id,
+        _keys: &AvailableCollections,
+        _plan: &GetPlan,
+    ) -> Self::Domain {
+        // A get operator yields physically monotonic output iff the correspoding
+        // `Plan::Get` is on a local or global ID that is known to provide physically
+        // monotonic input. The way this becomes know is through the interpreter itself
+        // for non-recursive local IDs or through configuration for the global IDs of
+        // monotonic sources and indexes. Recursive local IDs are always assumed to
+        // break physical monotonicity.
+        // TODO(vmarcos): Consider in the future if we can ascertain whether the
+        // restrictions on recursive local IDs can be relaxed to take into account only
+        // the interpreter judgement directly.
+        PhysicallyMonotonic(match id {
+            Id::Local(id) => ctx
+                .bindings
+                .get(id)
+                .map_or(false, |entry| !entry.is_rec && entry.value.0),
+            Id::Global(id) => self.monotonic_ids.contains(id),
+        })
+    }
+
+    fn mfp(
+        &self,
+        _ctx: &Context<Self::Domain>,
+        input: Self::Domain,
+        _mfp: &MapFilterProject,
+        _input_key_val: &Option<(Vec<MirScalarExpr>, Option<Row>)>,
+    ) -> Self::Domain {
+        // In a single-time context, we just propagate the monotonicity
+        // status of the input
+        input
+    }
+
+    fn flat_map(
+        &self,
+        _ctx: &Context<Self::Domain>,
+        input: Self::Domain,
+        _func: &TableFunc,
+        _exprs: &Vec<MirScalarExpr>,
+        _mfp: &MapFilterProject,
+        _input_key: &Option<Vec<MirScalarExpr>>,
+    ) -> Self::Domain {
+        // In a single-time context, we just propagate the monotonicity
+        // status of the input
+        input
+    }
+
+    fn join(
+        &self,
+        _ctx: &Context<Self::Domain>,
+        inputs: Vec<Self::Domain>,
+        _plan: &JoinPlan,
+    ) -> Self::Domain {
+        // When we see a join, we must consider that the inputs could have
+        // been `Plan::Get`s on arrangements. These are not in general safe
```

Comment:
> When we see a join, we must consider that the inputs could have been `Plan::Get`s on arrangements.

I think this is related to my comment on the docs above. I don't understand why this assumption is not sufficient.

Reply:
I added a couple of explanatory sentences now in the design doc. The issue is that when we import an index through `import_frontier_core`, we lose the guarantee that a physically monotonic single-time stream will be presented even though the stream originates from an arrangement.